### PR TITLE
change ALL_DIMS to equal ellipsis

### PR DESCRIPTION
**Pull request by max-sixty**
- [x] Closes https://github.com/pydata/xarray/issues/3414
 - [x] Passes `black . && mypy . && flake8`
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

This is a more drastic version of the suggested change, making `xr.ALL_DIMS` equal the ellipsis, `...`. 

The only downside I can see is that printing `xr.ALL_DIMS` will print `...`. The upside is that we get compat 'for free'

### CODE REVIEW

**Comment by crusaderky** [comment]
Tests broken by #3409 - please stand by...

**Comment by max-sixty** [comment]
Should we advertise `...` as the recommended way to supply 'all dims', rather than `xr.ALL_DIMS`? (if we're not sure, we could leave as currently proposed and change later)

**Comment by crusaderky** [comment]
@max-sixty I'm in favour of ... as the preferred syntax

**Comment by max-sixty** [comment]
Ready for review / merge

**Comment by max-sixty** [comment]
Hmmm, still getting cftime issues with the build: `ERROR: Command errored out with exit status 128: git fetch -q https://github.com/Unidata/cftime.git refs/pull/127/merge Check the logs for full command output.`
https://dev.azure.com/xarray/xarray/_build/results?buildId=1208&view=logs

**Comment by dcherian** [comment]
`cftime` has merged the PR with the fix so we should just switch back to their master branch

**Comment by shoyer** [comment]
build failure should be fixed by https://github.com/pydata/xarray/pull/3439

**Comment by max-sixty** [comment]
Any final comments before merge?

File: xarray/core/groupby.py [hunk 1]
```diff
@@ -895,9 +895,9 @@ def reduce(self, func, dim=None, keep_attrs=None, **kwargs):
         def reduce_dataset(ds):
             return ds.reduce(func, dim, keep_attrs, **kwargs)
 
-        if dim is not ALL_DIMS and dim not in self.dims:
+        if dim is not ... and dim not in self.dims:
             raise ValueError(
-                "cannot reduce over dimension %r. expected either xarray.ALL_DIMS to reduce over all dimensions or one or more of %r."
+                "cannot reduce over dimension %r. expected either `...` to reduce over all dimensions or one or more of %r."
```
Comment (dcherian) [inline xarray/core/groupby.py]:
```suggestion
                "cannot reduce over dimension %r. expected either '...' to reduce over all dimensions or one or more of %r."
```

File: doc/groupby.rst [hunk 2]
```diff
@@ -116,7 +116,7 @@ dimensions *other than* the provided one:
 
 .. ipython:: python
 
-    ds.groupby('x').std(xr.ALL_DIMS)
+    ds.groupby('x').std(...)
```
Comment (dcherian) [inline doc/groupby.rst]: Describe `...` in a tip box here?

File: xarray/core/groupby.py [hunk 3]
```diff
@@ -794,9 +794,9 @@ def reduce(
         if keep_attrs is None:
             keep_attrs = _get_keep_attrs(default=False)
 
-        if dim is not ALL_DIMS and dim not in self.dims:
+        if dim is not ... and dim not in self.dims:
             raise ValueError(
-                "cannot reduce over dimension %r. expected either xarray.ALL_DIMS to reduce over all dimensions or one or more of %r."
+                "cannot reduce over dimension %r. expected either `...` to reduce over all dimensions or one or more of %r."
```
Comment (dcherian) [inline xarray/core/groupby.py]:
```suggestion
                "cannot reduce over dimension %r. expected either '...' to reduce over all dimensions or one or more of %r."
```

File: doc/groupby.rst (hunk 2 above)
Comment (max-sixty) [reply doc/groupby.rst]: Good idea, done

Comment (max-sixty) [reply doc/groupby.rst]: Good idea; done

**Comment by dcherian** [comment]
Thanks @max-sixty !

**Pull request merged**